### PR TITLE
Integrating batchcorp/rabbit for relay reconnect support

### LINKED ISSUE #28: Update rabbitmq lib to support reconnect
Current rabbitmq implementation freaks out on disconnect. Update it to support auto-reconnects.

### PR DESCRIPTION
**Pull request by blinktag**
Replacing rabbitmq consumption code with batchcorp/rabbit

Resolves #28

### CODE REVIEW

**Review by dselans — CHANGES_REQUESTED** [review]
`Consume` should be triggered outside the for loop & select. Rest is nitpicks.

**Comment by dselans** [comment]
Let's also update `read` and `write` to use the new lib.

File: backends/rabbitmq/relay.go [hunk 1]
```diff
@@ -91,41 +93,43 @@ func validateRelayOptions(opts *cli.Options) error {
...
+	rmq, err := rabbit.New(&rabbit.Options{
+		URL:          r.Options.Rabbit.Address,
+		QueueName:    r.Options.Rabbit.ReadQueue,
+		ExchangeName: r.Options.Rabbit.Exchange,
+		RoutingKey:   r.Options.Rabbit.RoutingKey,
+	})
 
 	if err != nil {
-		return errors.Wrap(err, "unable to create initial consume channel")
+		return errors.Wrap(err, "unable to start rabbitmq consumer")
```
[the remark this answers is not in the document]
Comment (blinktag) [reply backends/rabbitmq/relay.go]: Fixed

File: backends/rabbitmq/relay.go [hunk 2]
```diff
@@ -91,41 +93,43 @@ func validateRelayOptions(opts *cli.Options) error {
...
-				Value:   &msg,
-				Options: &types.RelayMessageOptions{},
-			}
-		case <-r.DefaultContext.Done():
-			r.log.Warning("received notice to quit loop")
-			r.Looper.Quit()
+		case errRabbit := <-errCh:
+			r.log.Errorf("error receiving message from rabbit %s", errRabbit)
+		default:
+			rmq.Consume(ctx, errCh, func(msg amqp.Delivery) error {
```
[the remark this answers is not in the document]
Comment (blinktag) [reply backends/rabbitmq/relay.go]: Fixed

File: backends/rabbitmq/relay.go [hunk 3]
```diff
@@ -91,41 +93,43 @@ func validateRelayOptions(opts *cli.Options) error {
...
-
-			r.RelayCh <- &types.RelayMessage{
-				Value:   &msg,
-				Options: &types.RelayMessageOptions{},
-			}
-		case <-r.DefaultContext.Done():
-			r.log.Warning("received notice to quit loop")
-			r.Looper.Quit()
+		case errRabbit := <-errCh:
+			r.log.Errorf("error receiving message from rabbit %s", errRabbit)
```
[the remark this answers is not in the document]
Comment (blinktag) [reply backends/rabbitmq/relay.go]: Fixed

File: backends/rabbitmq/rabbitmq_test.go [hunk 4]
```diff
@@ -24,6 +30,7 @@ func init() {
 }
 
 var _ = Describe("Shared", func() {
+
 	Context("connect", func() {
```
Comment (dselans) [inline backends/rabbitmq/rabbitmq_test.go:34]: Let's name the contexts or describe's for the func's they're testing. I think this is `Write`, no?

EDIT: Read/Write

File: backends/rabbitmq/read.go [hunk 5]
```diff
@@ -34,38 +35,26 @@ func Read(opts *cli.Options) error {
...
-	}
+	r, err := New(opts, md)
 
-	r := &RabbitMQ{
-		Options: opts,
-		Channel: ch,
-		MsgDesc: md,
-		log:     logrus.WithField("pkg", "rabbitmq/read.go"),
+	if err != nil {
+		return err
```
Comment (dselans) [inline backends/rabbitmq/read.go:41]: Let's wrap this in `errors.Wrap(err, "informative message about where this occurred")`

File: backends/rabbitmq/read.go [hunk 6]
```diff
@@ -34,38 +35,26 @@ func Read(opts *cli.Options) error {
...
-	}
+	errCh := make(chan *rabbit.ConsumeError)
+	ctx, cancel := context.WithCancel(context.Background())
 
-	lineNumber := 1
+	go r.Consumer.Consume(ctx, errCh, func(msg amqp.Delivery) error {
 
-	for {
-		msg := <-msgChan
+		lineNumber := 1
```
Comment (dselans) [inline backends/rabbitmq/read.go:57]: Is this going to be 1 always?

File: backends/rabbitmq/read.go [hunk 7]
```diff
@@ -99,7 +88,7 @@ func (r *RabbitMQ) Read() error {
 			}
 
 			printer.Error(fmt.Sprintf("unable to complete conversion for message: %s", convertErr))
-			continue
+			return convertErr
```
Comment (dselans) [inline backends/rabbitmq/read.go:91]: Wrap this in `errors.Wrap()`

File: backends/rabbitmq/relay.go [hunk 8]
```diff
@@ -91,41 +86,43 @@ func validateRelayOptions(opts *cli.Options) error {
...
-				Options: &types.RelayMessageOptions{},
-			}
-		case <-r.DefaultContext.Done():
-			r.log.Warning("received notice to quit loop")
-			r.Looper.Quit()
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go rmq.Consume(ctx, errCh, func(msg amqp.Delivery) error {
+		r.log.Debug(fmt.Printf("Writing RabbitMQ message to relay channel: %+v", msg))
+		r.RelayCh <- &types.RelayMessage{
```
Comment (dselans) [inline backends/rabbitmq/relay.go:112]: What a pretty interface 😍

File: backends/rabbitmq/write.go [hunk 9]
```diff
@@ -38,32 +36,26 @@ func Write(opts *cli.Options) error {
...
-	}
+	r, err := New(opts, md)
 
-	r := &RabbitMQ{
-		Options: opts,
-		Channel: ch,
-		MsgDesc: md,
-		log:     logrus.WithField("pkg", "rabbitmq/write.go"),
+	if err != nil {
+		return err
```
Comment (dselans) [inline backends/rabbitmq/write.go:42]: `errors.Wrap()`

**Review by dselans — CHANGES_REQUESTED** [review]
Minor notes - great job!

File: backends/rabbitmq/read.go (hunk 7 above)
Comment (blinktag) [reply backends/rabbitmq/read.go]: Fixed

File: backends/rabbitmq/write.go (hunk 9 above)
Comment (blinktag) [reply backends/rabbitmq/write.go]: Fixed

File: backends/rabbitmq/rabbitmq_test.go (hunk 4 above)
Comment (blinktag) [reply backends/rabbitmq/rabbitmq_test.go]: Fixed

File: backends/rabbitmq/read.go (hunk 6 above)
Comment (blinktag) [reply backends/rabbitmq/read.go]: Fixed

**Review by dselans — APPROVED** [review]
🎉

**Pull request merged**